### PR TITLE
Financial Assurance Tab Partners

### PR DESCRIPTION
**Pull request by marko911**
connects https://app.clubhouse.io/unicefetools/story/4241/pmp-changes-for-hact

### CODE REVIEW

File: pmp/pages/page-partners.html [hunk 1]
```diff
@@ -401,9 +413,15 @@ <h1 main-title>
...
         }
-        let partnerDetails = this.$$('#partnerDetails');
-        if (partnerDetails && partnerDetails._validatePartner()) {
-          var partnerChanges = this._cleanUpdateData(this.partner);
+
+        let pageToValidate = this.partnerPage==='details' ?
+          this.$$('#partnerDetails')
+          : this.partnerPage === 'financial-assurance' ?
+            this.$$('#financialAssurance')
+            : null;
```
Comment (daniel-tabarcea) [inline pmp/pages/page-partners.html]: @marko911 Can you please rewrite this so it uses multiple Ifs instead of a single inline condition? It's hard to follow, and we're trying to make everything as easy to read as we can

File: pmp/app-elements/partners/partner-details.html [hunk 2]
```diff
@@ -202,11 +198,6 @@
             </paper-input>
           </div>
         </div>
-
-        <!-- Assessments section -->
-        <div class="assessments">
-          <assessments-items id="assessmentsList" data-items="{{partner.assessments}}" edit-mode="[[editMode]]"></assessments-items>
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-details.html]: remove `assessments-items.html` since it  was moved from this element

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+        height: 72px;
+        box-sizing: border-box;
+      }
+
+      .overview-header {
+        background-color: var(--medium-theme-background-color, #eeeeee);
+      }
+
+      .risk-rating {
+        @apply(--risk-rating-color);
```
Comment (daniel-tabarcea) [inline pmp/app-elements/partners/partner-financial-assurance.html]: should be @apply --risk-rating-color;

File: pmp/app-elements/partners/partner-details.html [hunk 4]
```diff
@@ -299,15 +291,8 @@
...
-        _validatePartner: function() {
-          var valid = true;
-          if (!this.$.assessmentsList.validate()) {
-            valid = false;
-          }
-          if (!this.$.staffMembersList.validate()) {
-            valid = false;
-          }
-          return valid;
+        _validatPage: function() {
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-details.html]: `_validatePageData` or `_validatePartnerData` would be a better name

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+        if (!this.pageSize) {
+          this.set('pageSize', 5);
+          this.set('pageNumber', 1);
+        }else{
+          this._paginate();
+        }
+        this._addBasisFromEngagements(engagements);
+      },
+
+      _displayType(type){ return this.TYPES[type]},
```
Comment (daniel-tabarcea) [inline pmp/app-elements/partners/partner-financial-assurance.html]: If you have the time, please fix the linting issues

File: pmp/pages/page-partners.html [hunk 6]
```diff
@@ -401,9 +413,15 @@ <h1 main-title>
         if (!this._hasEditPermissions(this.permissions)) {
           return;
         }
-        let partnerDetails = this.$$('#partnerDetails');
-        if (partnerDetails && partnerDetails._validatePartner()) {
-          var partnerChanges = this._cleanUpdateData(this.partner);
+
+        let pageToValidate = this.partnerPage==='details' ?
```
Comment (adi130987) [inline pmp/pages/page-partners.html]: Please change this to something more readable

File: pmp/pages/page-partners.html (hunk 1 above)
Comment (marko911) [reply pmp/pages/page-partners.html]: I find that a ternary is much easier to read than a bunch of nested ifs. But I can change it I don't really care.

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 5 above)
Comment (marko911) [reply pmp/app-elements/partners/partner-financial-assurance.html]: we dont have the same linter so I dont know which issues those are

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 3 above)
Comment (marko911) [reply pmp/app-elements/partners/partner-financial-assurance.html]: both works

Comment (adi130987) [reply pmp/app-elements/partners/partner-financial-assurance.html]: `@apply --risk-rating-color;` is the new way of applying mixins (Polymer 2), but works in Polymer 1.x too

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 5 above)
Comment (adi130987) [reply pmp/app-elements/partners/partner-financial-assurance.html]: make sure you have `gulp-cli` installed, then run: `gulp lint --max_old_space_size=4096`.
There are other linting issues that need to be solved. 
Ignore `poly lint` issues like: `<partner-financial-assurance> is undefined.` (because of layzi load functionality) and `...bound to the attribute 'data-items' should use $= instead of =.` elements that use repeatable data set behavior

File: bower.json [hunk 7]
```diff
@@ -70,6 +70,8 @@
     "lodash": "^4.17.4",
     "paper-checkbox": "2.0.0",
     "paper-radio-button": "2.0.0",
-    "webcomponentsjs": "^0.7.24"
+    "webcomponentsjs": "^0.7.24",
+    "etools-ajax": "^2.0.3",
```
Comment (adi130987) [inline bower.json]: We need to fix `etools-ajax` and `etools-loading` resolutions.

File: pmp/app-elements/partners/partner-details.html [hunk 8]
```diff
@@ -295,15 +286,8 @@
...
-        _validatePartner: function() {
-          var valid = true;
-          if (!this.$.assessmentsList.validate()) {
-            valid = false;
-          }
-          if (!this.$.staffMembersList.validate()) {
-            valid = false;
-          }
-          return valid;
+        _validatPageData: function() {
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-details.html]: Typo issue in `_validatPageData`, it should be  like in `page-partners`, `pageToValidate._validatePageData()`.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 9]
```diff
@@ -0,0 +1,642 @@
+<link rel="import" href="../../../bower_components/polymer/polymer.html">
+<link rel="import" href="../../../bower_components/iron-flex-layout/iron-flex-layout.html">
+<link rel="import" href="../../../bower_components/iron-flex-layout/iron-flex-layout-classes.html">
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: We do not use `iron-flex-layout-classes`, if you don't have everything you need in `grid-layout-styles`, if not use `iron-flex-layout` mixins on you template elements, you're gonna use just a few of them.

For example: `layout horizontal` is already defined in app grid styles: `layout-horizontal` and instead of center class you can use `@apply --layout-center` on `.overview-header` or `overview-row`.

Comment (adi130987) [reply pmp/app-elements/partners/partner-financial-assurance.html]: `.row-h` already has applied `layout` and `horizontal` classes style, avoid `<div class="panel-row-tall overview-header overview-row layout horizontal center row-h">` ... too many classes that have the same css.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+        color: var(--primary-text-color);
+      }
+
+    </style>
+
+    <etools-content-panel
+        panel-title="Overview"
+        class="content-section"
+    >
+      <div class="panel-row-tall overview-header overview-row layout horizontal center row-h">
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: .row-h already has applied layout and horizontal classes style, avoid <div class="panel-row-tall overview-header overview-row layout horizontal center row-h"> ... too many classes that have the same css.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+      .view-only {
+        background-color: #FAFAFA;
+      }
+
+      etools-data-table-column, .table-title {
+        font-weight :500;
+      }
+
+      .panel-row-tall {
+        height: 56px;
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: using this height on header row forces the column header title on 2 lines to overlap header container
screenshot from 2018-03-05 14-30-45

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+        cursor: pointer;
+        margin-left: -8px;
+        padding-left: 24px;
+      }
+
+      .vision {
+        position: relative;
+        width: 66.66667%;
+        font-size: 16px;
+        border: 2px solid rgba(0,97,233,0.38);
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: Either we use a similar color from `app-theme` (to avoid adding new colors and keep theme color consistency), or define a new css variable in app-theme for this rgba

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+        </etools-data-table-column>
+      </div>
+      <div class="row-h center overview-row">
+        <div class="layout horizontal center vision">
+          <div class="from-vision">from VISION</div>
+          <div class="layout horizontal col-3">
+            <div class="risk-rating"></div>
+            [[partner.rating]]
+          </div>
+          <div class="col-6">[[partner.basis_for_risk_rating]]</div>
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: use `col col-6`. `col` class will ensure the correct space(`24px`) between these elements.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+          SPOT CHECKS <br> Done / MR
+        </etools-data-table-column>
+        <etools-data-table-column class="col-2 col">
+          AUDIT <br> Done / MR
+        </etools-data-table-column>
+      </div>
+      <div class="row-h center overview-row">
+        <div class="layout horizontal center vision">
+          <div class="from-vision">from VISION</div>
+          <div class="layout horizontal col-3">
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: instead of `layout horizontal col-3` use `col col-3`... `col` has `--layout-horizontal` applied in app grid styles

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 15]
```diff
@@ -0,0 +1,642 @@
...
+            $[[displayCurrencyAmount(partner.total_ct_cy, '0', 0)]]
+          </div>
+        </div>
+        <div class="col-2 col layout horizontal center hact-values">
+          <strong>
+            <span class="green">[[partner.hact_values.spot_checks.completed.total]] </span>
+            / [[partner.hact_min_requirements.spot_checks]]
+          </strong>
+        </div>
+        <div class="col-2 col layout horizontal center hact-values">
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: instead of `center` I think you ca use `center-align` from app grid styles

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 16]
```diff
@@ -0,0 +1,642 @@
...
+    </etools-content-panel>
+
+    <etools-content-panel
+        panel-title="[[_getYear()]] Financial Assurance Planning"
+        class="content-section"
+    >
+      <div class="planning-wrapper">
+        <div class="layout horizontal">
+          <div class="table-main col-6 layout vertical spot-checks">
+            <div class="table-main panel-row-tall row-h panel-table-row layout horizontal center view-only">
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: `table-main panel-row-tall row-h panel-table-row layout horizontal center view-only` too many classes

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 17]
```diff
@@ -0,0 +1,642 @@
...
+            'audit': 'Audit',
+            'ma': 'Micro Assessment',
+            'sc': 'Spot Check',
+            'sa': 'Special Audit'
+          }
+        },
+        RISKRATINGCOLORS:{
+          type: Object,
+          value: {
+            'Not Required': '#D8D8D8',
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: Use color variable prefixed with `--risk-rating` from `app-theme`, update colors if needed.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 18]
```diff
@@ -0,0 +1,642 @@
...
+            csrf: true,
+          }
+        },
+        partner: {
+          type: Object,
+          observer: '_partnerReceived'
+        },
+        TYPES: {
+          type: Object,
+          value: {
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: Don't we have some endpoint that returns this data?
Assessments types are loaded from `/api/v2/dropdowns/static/` endpoint response than are stored in app redux store. We should have these types returned by this endpoint too. Maybe @robertavram or @ntrncic can help us with this.

File: pmp/app-elements/partners/partner-financial-assurance.html [hunk 19]
```diff
@@ -0,0 +1,642 @@
...
+        /**
+         * Disable loading message for details tab elements load,
+         * triggered by parent element on stamp or by tap event on tabs
+         */
+        this.fire('global-loading', {active: false, loadingSource: 'partners-page'});
+        this.fire('tab-content-attached');
+      },
+
+      _init(engagements){
+        this.set('totalEngagements', engagements);
```
Comment (adi130987) [inline pmp/app-elements/partners/partner-financial-assurance.html]: `totalEngagements` makes me think of the total number of engagements. For better readability we could rename this (and add to properties section) to `allEngagements` or `partnerEngagements`.

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 18 above)
Comment (marko911) [reply pmp/app-elements/partners/partner-financial-assurance.html]: You would rather call an endpoint and store this in redux? I don't think this mapping is used anywhere else. @adi130987 @robertavram

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 17 above)
Comment (marko911) [reply pmp/app-elements/partners/partner-financial-assurance.html]: How can you make that work in polymer ? Change a style based on a variable in js?

File: pmp/app-elements/partners/partner-financial-assurance.html (hunk 18 above)
Comment (adi130987) [reply pmp/app-elements/partners/partner-financial-assurance.html]: I will talk with Rob about this. In PMP all static data is loaded from API and cached(at first load) stored in app redux store and made available to entire app/any element.
